### PR TITLE
LockWatchManager#isEnabled.

### PR DESCRIPTION
**Pull request by jkozlowski**
**Goals (and why)**:

**Implementation Description (bullets)**:

**Testing (What was existing testing like?  What have you done to improve it?)**:

**Concerns (what feedback would you like?)**:

**Where should we start reviewing?**:

**Priority (whenever / two weeks / yesterday)**:

### CODE REVIEW

File: changelog/@unreleased/pr-5040.v2.yml [hunk 1]
```diff
@@ -0,0 +1,24 @@
+type: improvement
+improvement:
+  description: |-
```
Comment (Jolyon-S) [inline changelog/@unreleased/pr-5040.v2.yml:3]: Fix this changelog, please!

File: lock-api/src/main/java/com/palantir/lock/watch/LockWatchEventCache.java [hunk 2]
```diff
@@ -21,6 +21,11 @@
 import java.util.Set;
 
 public interface LockWatchEventCache {
+
+    default boolean isEnabled() {
+        return true;
+    }
```
Comment (Jolyon-S) [inline lock-api/src/main/java/com/palantir/lock/watch/LockWatchEventCache.java:27]: This feels wrong - this should call down to `LockWatchEventCacheImpl` or `NoOpLockWatchEventCache`.

File: atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/keyvalue/api/watch/ResilientLockWatchEventCacheTest.java [hunk 3]
```diff
@@ -54,6 +56,21 @@ public void before() {
...
     }
 
+    @Test
+    public void testCanDelegateIsEnabled() {
+        assertThat(proxyCache.isEnabled()).isTrue();
+
+        RuntimeException runtimeException = new RuntimeException();
+        when(defaultCache.getCommitUpdate(anyLong())).thenThrow(runtimeException);
+        assertThatThrownBy(() -> proxyCache.getCommitUpdate(0L)).hasCause(runtimeException)
+                .isExactlyInstanceOf(TransactionLockWatchFailedException.class);
```
Comment (Jolyon-S) [inline atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/keyvalue/api/watch/ResilientLockWatchEventCacheTest.java:66]: do you mind expanding this test with `hasMessage` as there are a number of exceptions under that class (I know I am guilty of this too, but in an effort to make it clearer what we are testing, this should be done).

File: atlasdb-api/src/main/java/com/palantir/atlasdb/keyvalue/api/watch/LockWatchManager.java [hunk 4]
```diff
@@ -34,6 +34,10 @@
 
     // These methods are hidden on purpose as they should not be generally available, only for brave souls!
 
+    boolean isEnabled() {
+        return true;
+    }
```
Comment (Jolyon-S) [inline atlasdb-api/src/main/java/com/palantir/atlasdb/keyvalue/api/watch/LockWatchManager.java:39]: Why is this necessary? Also, if we're going to default to anything, wouldn't it be better to be false? I still think this should just be:
`boolean isEnabled;`

File: lock-api/src/main/java/com/palantir/lock/watch/LockWatchEventCache.java (hunk 2 above)
Comment (Jolyon-S) [reply lock-api/src/main/java/com/palantir/lock/watch/LockWatchEventCache.java:27]: Oh I see - you went with an approach entirely through the proxy. I'm not sure I like that really - surely the logic is 100x cleaner just by doing what you originally said, i.e. having `LWECI` return true and `NOLWEC` return false?

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/keyvalue/api/watch/ResilientLockWatchEventCache.java [hunk 5]
```diff
@@ -34,6 +34,15 @@
...
     private static final Logger log = LoggerFactory.getLogger(ResilientLockWatchEventCache.class);
+    private static final Method IS_ENABLED;
+
+    static {
+        try {
+            IS_ENABLED = LockWatchEventCache.class.getMethod("isEnabled");
+        } catch (NoSuchMethodException e) {
+            throw new RuntimeException("Did someone rename this method?", e);
+        }
+    }
```
Comment (Jolyon-S) [inline atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/keyvalue/api/watch/ResilientLockWatchEventCache.java:45]: I don't really like this - I much prefer just implementing in LWECI and NOLWEC as I feel it is less error prone, and doing this kind of proxy-wrangling is abusing the whole point of it IMO.

**Review by Jolyon-S — CHANGES_REQUESTED** [review]
I'm not really sure I like the approach of abusing the proxy like this - can we not just go with the first thing you suggested, which is just have the classes implement `isEnabled` to true or false?

File: lock-api/src/main/java/com/palantir/lock/watch/LockWatchEventCache.java (hunk 2 above)
Comment (jkozlowski) [reply lock-api/src/main/java/com/palantir/lock/watch/LockWatchEventCache.java:27]: Discussed offline

File: atlasdb-api/src/main/java/com/palantir/atlasdb/keyvalue/api/watch/LockWatchManager.java (hunk 4 above)
Comment (jkozlowski) [reply atlasdb-api/src/main/java/com/palantir/atlasdb/keyvalue/api/watch/LockWatchManager.java:39]: I don’t think it matters, I just went with this approach.

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/keyvalue/api/watch/ResilientLockWatchEventCache.java (hunk 5 above)
Comment (jkozlowski) [reply atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/keyvalue/api/watch/ResilientLockWatchEventCache.java:45]: Discussed offline

File: changelog/@unreleased/pr-5040.v2.yml (hunk 1 above)
Comment (jkozlowski) [reply changelog/@unreleased/pr-5040.v2.yml:3]: Done

File: atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/keyvalue/api/watch/ResilientLockWatchEventCacheTest.java (hunk 3 above)
Comment (jkozlowski) [reply atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/keyvalue/api/watch/ResilientLockWatchEventCacheTest.java:70]: Nope, ResilientLockWatchEventCache has just one from what I can see?

File: atlasdb-api/src/main/java/com/palantir/atlasdb/keyvalue/api/watch/LockWatchManager.java (hunk 4 above)
Comment (Jolyon-S) [reply atlasdb-api/src/main/java/com/palantir/atlasdb/keyvalue/api/watch/LockWatchManager.java:39]: I still think we should not set this - I'd rather never, ever have to debug another bug like that auto-delegate, and while I know this is very different, it's just less error-prone to not set a default.

Comment (Jolyon-S) [reply atlasdb-api/src/main/java/com/palantir/atlasdb/keyvalue/api/watch/LockWatchManager.java:39]: (I know you have a test for it, but still).

File: atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/keyvalue/api/watch/ResilientLockWatchEventCacheTest.java (hunk 3 above)
Comment (Jolyon-S) [reply atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/keyvalue/api/watch/ResilientLockWatchEventCacheTest.java:70]: Ah yea, I see you're just throwing a generic exception instead of actually causing a legit exception inside LWECI.

**Review by Jolyon-S — APPROVED** [review]
Just one thing, then I think good to merge.

File: atlasdb-api/src/main/java/com/palantir/atlasdb/keyvalue/api/watch/LockWatchManager.java (hunk 4 above)
Comment (jkozlowski) [reply atlasdb-api/src/main/java/com/palantir/atlasdb/keyvalue/api/watch/LockWatchManager.java]: Done
